File: TagMaster/src/depollsoft/tagmaster/QueryModel.java
package depollsoft.tagmaster;

import depollsoft.lib.binding.ObservableCollection;
import depollsoft.lib.binding.TrackableField;
import depollsoft.lib.ui.ThreadSwitchContext;
import depollsoft.lib.util.Action;
import depollsoft.tagmaster.barbershop.Tag;
import depollsoft.tagmaster.barbershop.TagCollection;
import depollsoft.tagmaster.barbershop.TagQueryResult;
import depollsoft.tagmaster.barbershop.TagSortOptions;

public class QueryModel
{
   private TagQueryResult mostRecentResult;

   private TrackableField<Integer> maxResults = new TrackableField<Integer>();

   private TrackableField<String> statusText = new TrackableField<String>();

   private TrackableField<TagCollection> collection = new TrackableField<TagCollection>();

   private TrackableField<Boolean> hasMoreResults = new TrackableField<Boolean>();

   private TrackableField<Boolean> isLoading = new TrackableField<Boolean>();

   private TrackableField<String> query = new TrackableField<String>();

   private TrackableField<Integer> resultSetSize = new TrackableField<Integer>();

   private TrackableField<Integer> parts = new TrackableField<Integer>();

   private TrackableField<Boolean> hasLearningTracks = new TrackableField<Boolean>();

   private TrackableField<Boolean> hasSheetMusic = new TrackableField<Boolean>();

   private TrackableField<TagSortOptions> sortBy = new TrackableField<TagSortOptions>();

   private TrackableField<ObservableCollection<Tag>> tags = new TrackableField<ObservableCollection<Tag>>();

   private TrackableField<Double> minRating = new TrackableField<Double>();

   private TrackableField<Integer> minDownloads = new TrackableField<Integer>();

   public QueryModel()
   {
      this.mostRecentResult = new TagQueryResult();
      this.mostRecentResult.setStart(0);
      this.mostRecentResult.setCount(0);
      this.setIsLoading(false);
      this.setHasMoreResults(true);
      this.setResultSetSize(20);
      this.setMaxResults(50);
      this.setTags(new ObservableCollection<Tag>());
   }

   public void fetchResults(final ThreadSwitchContext context)
   {
      if (this.getIsLoading())
         return;
      if (!this.getHasMoreResults())
         return;
      this.setIsLoading(true);
      Tag.query(
            this.getQuery(),
            this.getResultSetSize(),
            this.mostRecentResult.getStart() + this.mostRecentResult.getCount(),
            this.getParts(), this.getHasLearningTracks(),
            this.getHasSheetMusic(), this.getCollection(), this.getSortBy(),
            this.getMinimumRating(), this.getMinimumDownloads()).continueWith(
            new Action<TagQueryResult>()
            {

               public void invoke(final TagQueryResult parameter)
               {
                  context.post(new Runnable()
                  {

                     public void run()
                     {
                        try
                        {
                           QueryModel.this.setStatusText(null);
                           QueryModel.this.mostRecentResult = parameter;
                           for (Tag t : parameter.getTags())
                              QueryModel.this.getTags().add(t);
                           if (QueryModel.this.mostRecentResult.getStart()
                                 + QueryModel.this.mostRecentResult.getCount() >= Math
                                 .min(QueryModel.this.mostRecentResult
                                       .getAvailable(), QueryModel.this
                                       .getMaxResults()))
                              QueryModel.this.setHasMoreResults(false);
                           else
                              QueryModel.this.setHasMoreResults(true);
                           if (parameter.getAvailable() == 0)
                              QueryModel.this
                                    .setStatusText("No tags could be found that matched your query.");
                        }
                        finally
                        {
                           QueryModel.this.setIsLoading(false);
                        }
                     }
                  });
               }
            }, new Action<Exception>()
            {

               public void invoke(final Exception parameter)
               {
                  context.post(new Runnable()
                  {

                     public void run()
                     {
                        QueryModel.this.setStatusText("An error has occurred: "
                              + parameter.getMessage());
                        QueryModel.this.setIsLoading(false);
                     }
                  });
               }
            });
   }

   public TagCollection getCollection()
   {
      return this.collection.getValue();
   }

   public Boolean getHasLearningTracks()
   {
      return this.hasLearningTracks.getValue();
   }

   public Boolean getHasMoreResults()
   {
      return this.hasMoreResults.getValue();
   }

   public Boolean getHasSheetMusic()
   {
      return this.hasSheetMusic.getValue();
   }

   public Boolean getIsLoading()
   {
      return this.isLoading.getValue();
   }

   public Integer getMaxResults()
   {
      return this.maxResults.getValue();
   }

   public Integer getMinimumDownloads()
   {
      return this.minDownloads.getValue();
   }

   public Double getMinimumRating()
   {
      return this.minRating.getValue();
   }

   public Integer getParts()
   {
      return this.parts.getValue();
   }

   public String getQuery()
   {
      return this.query.getValue();
   }

   public Integer getResultSetSize()
   {
      return this.resultSetSize.getValue();
   }

   public TagSortOptions getSortBy()
   {
      return this.sortBy.getValue();
   }

   public String getStatusText()
   {
      return this.statusText.getValue();
   }

   public ObservableCollection<Tag> getTags()
   {
      return this.tags.getValue();
   }

   public void setCollection(TagCollection value)
   {
      this.collection.setValue(value);
   }

   public void setHasLearningTracks(Boolean value)
   {
      this.hasLearningTracks.setValue(value);
   }

   public void setHasMoreResults(Boolean value)
   {
      this.hasMoreResults.setValue(value);
   }

   public void setHasSheetMusic(Boolean value)
   {
      this.hasSheetMusic.setValue(value);
   }

   public void setIsLoading(Boolean value)
   {
      this.isLoading.setValue(value);
   }

   public void setMaxResults(Integer value)
   {
      this.maxResults.setValue(value);
   }

   public void setMinimumDownloads(Integer value)
   {
      this.minDownloads.setValue(value);
   }

   public void setMinimumRating(Double value)
   {
      this.minRating.setValue(value);
   }

   public void setParts(Integer value)
   {
      this.parts.setValue(value);
   }

   public void setQuery(String value)
   {
      this.query.setValue(value);
   }

   public void setResultSetSize(Integer value)
   {
      this.resultSetSize.setValue(value);
   }

   public void setSortBy(TagSortOptions value)
   {
      this.sortBy.setValue(value);
   }

   public void setStatusText(String value)
   {
      this.statusText.setValue(value);
   }

   public void setTags(ObservableCollection<Tag> value)
   {
      this.tags.setValue(value);
   }
}

File: Windows Phone 7/BarbershopTags/BarbershopTags/QueryModel.cs
using System;
using System.Collections.Generic;
using System.Collections.ObjectModel;
using System.ComponentModel;
using System.Windows;
using System.Windows.Input;
using BarbershopTags.Barbershop;
using SLaB.Utilities;
using System.Threading;

namespace BarbershopTags
{
    public class QueryModel : INotifyPropertyChanged
    {
        private TagQueryResult _MostRecentResult;
        private ObservableCollection<Tag> _TagsLoaded;
        public QueryModel()
        {
            _MostRecentResult = new TagQueryResult { Start = 0, Count = 0 };
            _HasMoreResults = true;
            _TagsLoaded = new ObservableCollection<Tag>();
            ResultSetSize = 10;
            MaxResults = 50;
            Tags = _TagsLoaded;
            FetchResultsCommand = new LambdaCommand<object>(obj =>
                {
                    FetchResults();
                });
        }


        private int _MaxResults;
        public int MaxResults
        {
            get
            {
                return _MaxResults;
            }
            set
            {
                if (!EqualityComparer<int>.Default.Equals(_MaxResults, value))
                {
                    _MaxResults = value;
                    PropertyChanged.Raise(this, new PropertyChangedEventArgs("MaxResults"));
                }
            }
        }

        public virtual void FetchResults()
        {
            if (IsLoading)
                return;
            IsLoading = true;
            UiUtilities.ExecuteOnUiThread(() =>
            {
                Tag.Query(Query, ResultSetSize, _MostRecentResult.Start + _MostRecentResult.Count, Parts, HasLearningTracks, HasSheetMusic, Collection, SortBy).Continue(result =>
                {
                    Thread.Sleep(100);
                    UiUtilities.ExecuteOnUiThread(() =>
                    {
                        try
                        {
                            StatusText = null;
                            _MostRecentResult = result;
                            foreach (var tag in result.Tags)
                                _TagsLoaded.Add(tag);
                            if (_MostRecentResult.Start + _MostRecentResult.Count >= Math.Min(_MostRecentResult.Available, MaxResults))
                                HasMoreResults = false;
                            else
                                HasMoreResults = true;
                            if (result.Available == 0)
                                StatusText = "No tags could be found that matched your query.";
                        }
                        finally
                        {
                            IsLoading = false;
                        }
                    });
                },
                    err =>
                    {
                        UiUtilities.ExecuteOnUiThread(() =>
                        {
                            StatusText = "An error has occurred: " + err.Message;
                            IsLoading = false;
                        });
                    });
            });
        }

        public ICommand FetchResultsCommand
        {
            get;
            private set;
        }


        private string _StatusText;
        public string StatusText
        {
            get
            {
                return _StatusText;
            }
            set
            {
                if (!EqualityComparer<string>.Default.Equals(_StatusText, value))
                {
                    _StatusText = value;
                    PropertyChanged.Raise(this, new PropertyChangedEventArgs("StatusText"));
                }
            }
        }

        private TagCollection? _Collection;
        [TypeConverter(typeof(TagCollectionTypeConverter))]
        public TagCollection? Collection
        {
            get
            {
                return _Collection;
            }
            set
            {
                if (!EqualityComparer<TagCollection?>.Default.Equals(_Collection, value))
                {
                    _Collection = value;
                    PropertyChanged.Raise(this, new PropertyChangedEventArgs("Collection"));
                }
            }
        }

        private bool _HasMoreResults;
        public bool HasMoreResults
        {
            get
            {
                return _HasMoreResults;
            }
            private set
            {
                if (!EqualityComparer<bool>.Default.Equals(_HasMoreResults, value))
                {
                    _HasMoreResults = value;
                    PropertyChanged.Raise(this, new PropertyChangedEventArgs("HasMoreResults"));
                }
            }
        }

        private bool _IsLoading;
        public bool IsLoading
        {
            get
            {
                return _IsLoading;
            }
            set
            {
                if (!EqualityComparer<bool>.Default.Equals(_IsLoading, value))
                {
                    _IsLoading = value;
                    PropertyChanged.Raise(this, new PropertyChangedEventArgs("IsLoading"));
                }
            }
        }

        private string _Query;
        public string Query
        {
            get
            {
                return _Query;
            }
            set
            {
                if (!EqualityComparer<string>.Default.Equals(_Query, value))
                {
                    _Query = value;
                    PropertyChanged.Raise(this, new PropertyChangedEventArgs("Query"));
                }
            }
        }


        private int _ResultSetSize;
        public int ResultSetSize
        {
            get
            {
                return _ResultSetSize;
            }
            set
            {
                if (!EqualityComparer<int>.Default.Equals(_ResultSetSize, value))
                {
                    _ResultSetSize = value;
                    PropertyChanged.Raise(this, new PropertyChangedEventArgs("ResultSetSize"));
                }
            }
        }


        private int? _Parts;
        [TypeConverter(typeof(NullableIntTypeConverter))]
        public int? Parts
        {
            get
            {
                return _Parts;
            }
            set
            {
                if (!EqualityComparer<int?>.Default.Equals(_Parts, value))
                {
                    _Parts = value;
                    PropertyChanged.Raise(this, new PropertyChangedEventArgs("Parts"));
                }
            }
        }


        private bool? _HasLearningTracks;
        [TypeConverter(typeof(NullableBoolConverter))]
        public bool? HasLearningTracks
        {
            get
            {
                return _HasLearningTracks;
            }
            set
            {
                if (!EqualityComparer<bool?>.Default.Equals(_HasLearningTracks, value))
                {
                    _HasLearningTracks = value;
                    PropertyChanged.Raise(this, new PropertyChangedEventArgs("HasLearningTracks"));
                }
            }
        }


        private bool? _HasSheetMusic;
        [TypeConverter(typeof(NullableBoolConverter))]
        public bool? HasSheetMusic
        {
            get
            {
                return _HasSheetMusic;
            }
            set
            {
                if (!EqualityComparer<bool?>.Default.Equals(_HasSheetMusic, value))
                {
                    _HasSheetMusic = value;
                    PropertyChanged.Raise(this, new PropertyChangedEventArgs("HasSheetMusic"));
                }
            }
        }


        private TagSortOptions _SortBy;
        public TagSortOptions SortBy
        {
            get
            {
                return _SortBy;
            }
            set
            {
                if (!EqualityComparer<TagSortOptions>.Default.Equals(_SortBy, value))
                {
                    _SortBy = value;
                    PropertyChanged.Raise(this, new PropertyChangedEventArgs("SortBy"));
                }
            }
        }

        private ObservableCollection<Tag> _Tags;
        public ObservableCollection<Tag> Tags
        {
            get
            {
                return _Tags;
            }
            private set
            {
                if (!EqualityComparer<ObservableCollection<Tag>>.Default.Equals(_Tags, value))
                {
                    _Tags = value;
                    PropertyChanged.Raise(this, new PropertyChangedEventArgs("Tags"));
                }
            }
        }
        public event PropertyChangedEventHandler PropertyChanged;
    }
}
